lib/irq.c

```diff
@@ -54,52 +54,87 @@ vfio_irq_idx_to_str(int index)
 }
 
 static long
-irqs_disable(vfu_ctx_t *vfu_ctx, uint32_t index)
+dev_get_irqinfo(vfu_ctx_t *vfu_ctx, struct vfio_irq_info *irq_info_in,
+                struct vfio_irq_info *irq_info_out)
+{
+    assert(vfu_ctx != NULL);
+    assert(irq_info_in != NULL);
+    assert(irq_info_out != NULL);
+
+    // Ensure provided argsz is sufficiently big and index is within bounds.
+    if ((irq_info_in->argsz < sizeof(struct vfio_irq_info)) ||
+        (irq_info_in->index >= VFU_DEV_NUM_IRQS)) {
+        vfu_log(vfu_ctx, LOG_DEBUG, "bad irq_info (size=%d index=%d)\n",
+                irq_info_in->argsz, irq_info_in->index);
+        return -EINVAL;
+    }
+
+    irq_info_out->count = vfu_ctx->irq_count[irq_info_in->index];
+    irq_info_out->flags = VFIO_IRQ_INFO_EVENTFD;
+
+    return 0;
+}
+
+int
+handle_device_get_irq_info(vfu_ctx_t *vfu_ctx, uint32_t size,
+                           struct vfio_irq_info *irq_info_in,
+                           struct vfio_irq_info *irq_info_out)
+{
+    assert(vfu_ctx != NULL);
+    assert(irq_info_in != NULL);
+    assert(irq_info_out != NULL);
+
+    if (size != sizeof(*irq_info_in) || size != irq_info_in->argsz) {
+        vfu_log(vfu_ctx, LOG_WARNING, "IRQ info size %d", size);
+        return -EINVAL;
+    }
+
+    return dev_get_irqinfo(vfu_ctx, irq_info_in, irq_info_out);
+}
+
+static long
+irqs_disable(vfu_ctx_t *vfu_ctx, uint32_t index, uint32_t start, uint32_t count)
 {
-    int *irq_efd = NULL;
-    uint32_t i;
+    size_t i;
+    int *efds;
 
     assert(vfu_ctx != NULL);
     assert(index < VFU_DEV_NUM_IRQS);
+    assert(start + count <= vfu_ctx->irq_count[index]);
```

Comment:
~~This should be a runtime error.~~ Ignore that, I see that `device_set_irqs_validate` ensures that.